File: init-service/app/services/cas_service.py
import hashlib
import logging

import boto3
from botocore.exceptions import BotoCoreError, ClientError, EndpointConnectionError

from app.config import settings


logger = logging.getLogger(__name__)
_storage_disabled = False
CAS_PREFIX = "yuvro/CAS"
READ_CHUNK_SIZE = 1024 * 1024

s3_client = boto3.client(
    "s3",
    aws_access_key_id=settings.aws_access_key_id or None,
    aws_secret_access_key=settings.aws_secret_access_key or None,
    endpoint_url=settings.s3_endpoint or None,
)


def hash_content(content: bytes) -> str:
    return hashlib.sha256(content).hexdigest()


def hash_file(file_path: str) -> tuple[str, int]:
    digest = hashlib.sha256()
    size_bytes = 0

    with open(file_path, "rb") as handle:
        for chunk in iter(lambda: handle.read(READ_CHUNK_SIZE), b""):
            digest.update(chunk)
            size_bytes += len(chunk)

    return digest.hexdigest(), size_bytes


def object_key(content_hash: str) -> str:
    return f"{CAS_PREFIX}/{content_hash[:2]}/{content_hash}"


def _disable_storage(reason: str) -> None:
    global _storage_disabled
    if _storage_disabled:
        return
    _storage_disabled = True
    logger.warning("Disabling CAS uploads for this process: %s", reason)


def upload_if_missing(content_hash: str, content: bytes) -> None:
    global _storage_disabled

    bucket = settings.s3_bucket
    if not bucket or _storage_disabled:
        return

    key = object_key(content_hash)
    try:
        s3_client.head_object(Bucket=bucket, Key=key)
        return
    except ClientError as exc:
        error_code = exc.response.get("Error", {}).get("Code", "")
        if error_code not in {"404", "NoSuchKey", "NotFound"}:
            raise
    except (EndpointConnectionError, BotoCoreError, OSError) as exc:
        _disable_storage(str(exc))
        return

    try:
        s3_client.put_object(Bucket=bucket, Key=key, Body=content)
    except (EndpointConnectionError, BotoCoreError, OSError) as exc:
        _disable_storage(str(exc))


def upload_file_if_missing(content_hash: str, file_path: str) -> None:
    global _storage_disabled

    bucket = settings.s3_bucket
    if not bucket or _storage_disabled:
        return

    key = object_key(content_hash)
    try:
        s3_client.head_object(Bucket=bucket, Key=key)
        return
    except ClientError as exc:
        error_code = exc.response.get("Error", {}).get("Code", "")
        if error_code not in {"404", "NoSuchKey", "NotFound"}:
            raise
    except (EndpointConnectionError, BotoCoreError, OSError) as exc:
        _disable_storage(str(exc))
        return

    try:
        with open(file_path, "rb") as handle:
            s3_client.put_object(Bucket=bucket, Key=key, Body=handle)
    except (EndpointConnectionError, BotoCoreError, OSError) as exc:
        _disable_storage(str(exc))

File: init-service/src/shared/cas-service.ts
import crypto from "crypto";
import fs from "fs";
import { HeadObjectCommand, PutObjectCommand } from "@aws-sdk/client-s3";
import { s3Client } from "./s3-service";
import config from "../config";
import logger from "../loaders/logger";

let storageDisabled = false;
const CAS_PREFIX = "yuvro/CAS";
const READ_CHUNK_SIZE = 1024 * 1024;

export function hashContent(content: Buffer | string): string {
  return crypto.createHash("sha256").update(content).digest("hex");
}

export function hashFile(filePath: string): Promise<[string, number]> {
  return new Promise((resolve, reject) => {
    const hash = crypto.createHash("sha256");
    let sizeBytes = 0;

    const stream = fs.createReadStream(filePath, { highWaterMark: READ_CHUNK_SIZE });

    stream.on("data", (chunk) => {
      hash.update(chunk);
      sizeBytes += chunk.length;
    });

    stream.on("end", () => {
      resolve([hash.digest("hex"), sizeBytes]);
    });

    stream.on("error", (err) => {
      reject(err);
    });
  });
}

export function objectKey(contentHash: string): string {
  return `${CAS_PREFIX}/${contentHash.substring(0, 2)}/${contentHash}`;
}

function disableStorage(reason: string): void {
  if (storageDisabled) return;
  storageDisabled = true;
  logger.warn(`Disabling CAS uploads for this process: ${reason}`);
}

export async function uploadIfMissing(contentHash: string, content: Buffer): Promise<void> {
  const bucket = config.S3_BUCKET;
  if (!bucket || storageDisabled) return;

  const key = objectKey(contentHash);

  // Check if exists
  try {
    const headCommand = new HeadObjectCommand({ Bucket: bucket, Key: key });
    await s3Client.send(headCommand);
    return;
  } catch (error: any) {
    const code = error.name || error.$metadata?.httpStatusCode;
    if (code !== "NotFound" && error.$metadata?.httpStatusCode !== 404) {
      // Endpoint connection error or other S3 exceptions
      disableStorage(error.message || String(error));
      return;
    }
  }

  // Upload since it is missing
  try {
    const putCommand = new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: content,
    });
    await s3Client.send(putCommand);
  } catch (error: any) {
    disableStorage(error.message || String(error));
  }
}

export async function uploadFileIfMissing(contentHash: string, filePath: string): Promise<void> {
  const bucket = config.S3_BUCKET;
  if (!bucket || storageDisabled) return;

  const key = objectKey(contentHash);

  // Check if exists
  try {
    const headCommand = new HeadObjectCommand({ Bucket: bucket, Key: key });
    await s3Client.send(headCommand);
    return;
  } catch (error: any) {
    const code = error.name || error.$metadata?.httpStatusCode;
    if (code !== "NotFound" && error.$metadata?.httpStatusCode !== 404) {
      disableStorage(error.message || String(error));
      return;
    }
  }

  // Upload since it is missing
  try {
    const fileStream = fs.createReadStream(filePath);
    const putCommand = new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: fileStream,
    });
    await s3Client.send(putCommand);
  } catch (error: any) {
    disableStorage(error.message || String(error));
  }
}
